### PR TITLE
save checkpoints and profiler output to the first logger

### LINKED ISSUE #12177: [RFC] Where to save checkpoints and profiler output in case of multiple loggers?
### Background 
Currently, in the case that the user does not specify a `dirpath` to `ModelCheckpoint` or a `dirpath` to the `Profiler`, we write our output to the Logger, when there is exactly one Logger. However, in the case of multiple loggers we just write to `default_root_dir`.

The exact priority followed for one logger:
1. if a `dirpath` is provided, write to it
2. if not, write to the Logger

The exact priority followed for multiple loggers (or no loggers):
1. if a `dirpath` is provided, write to it
2. if not, write to `default_root_dir`

The purpose of this issue is to make sure everyone is aware of this behavior, and to see if there is a way to improve it.

### Discussion Questions
I'd like to discuss two questions here.
1. Can we confirm that this is the behavior we want for when we have one logger?
2. What behavior do we want for when we have multiple loggers? I see a few options:
    A. Write to `default_root_dir` (keep current behavior)
    B. Write to first logger in the list of loggers
    C. Write to all loggers in the list of loggers 

cc @borda @awaelchli @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @ninginthecloud @otaj @carmocca @kaushikb11 @nbcsm @guotuofeng

### PR DESCRIPTION
**Pull request by tshu-w**
## What does this PR do?

Fixes #12177 save checkpoints and profiler output to the first logger, which is a blocker of #14188.

### Does your PR introduce any breaking changes? If yes, please list them.

Yes, as discussed in https://github.com/Lightning-AI/lightning/issues/12177.

I also delete `pytorch_lightning.utilities.logger._name` directly without marking them as deprecated since it is not used in another place and it doesn't break an API. 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/tests_pytorch/profilers/test_profiler.py [hunk 1]
```diff
@@ -463,17 +463,15 @@ def test_pytorch_profiler_multiple_loggers(tmpdir):
 
     def look_for_trace(trace_dir):
         """Determines if a directory contains a PyTorch trace."""
+        print(os.listdir(trace_dir))
```
Comment (awaelchli) [inline tests/tests_pytorch/profilers/test_profiler.py:466]: forgot to remove ? :)

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -592,14 +591,10 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
...
         elif trainer.loggers:
-            if len(trainer.loggers) == 1:
-                assert trainer.logger is not None
-                save_dir = trainer.logger.save_dir or trainer.default_root_dir
-            else:
-                save_dir = trainer.default_root_dir
-
-            name = _name(trainer.loggers)
-            version = _version(trainer.loggers)
+            assert trainer.loggers[0] is not None
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:594]: NIT: maybe with `elif len(trainer.loggers) > 0` the assert here can be avoided

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/profilers/test_profiler.py (hunk 1 above)
Comment (tshu-w) [reply tests/tests_pytorch/profilers/test_profiler.py]: Yeah 😅.

File: src/pytorch_lightning/callbacks/model_checkpoint.py (hunk 2 above)
Comment (tshu-w) [reply src/pytorch_lightning/callbacks/model_checkpoint.py]: That's right, modified.

I didn't think much about it at the time and thought to keep it as close to the original code as possible :)

File: src/pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -297,9 +297,8 @@ def __init__(
 
             logger: Logger (or iterable collection of loggers) for experiment tracking. A ``True`` value uses
                 the default ``TensorBoardLogger``. ``False`` will disable logging. If multiple loggers are
-                provided and the `save_dir` property of that logger is not set, local files (checkpoints,
-                profiler traces, etc.) are saved in ``default_root_dir`` rather than in the ``log_dir`` of any
-                of the individual loggers.
+                provided, local files (checkpoints, profiler traces, etc.) are saved in the ``log_dir`` of
+                the first logger.
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:301]: can we add same info here, too? https://pytorch-lightning.readthedocs.io/en/stable/common/checkpointing_intermediate.html?highlight=checkpointing#where

File: src/pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -2207,11 +2206,11 @@ def model(self, model: torch.nn.Module) -> None:
 
     @property
     def log_dir(self) -> Optional[str]:
-        if len(self.loggers) == 1:
-            if isinstance(self.logger, TensorBoardLogger):
-                dirpath = self.logger.log_dir
+        if len(self.loggers) > 0:
+            if isinstance(self.loggers[0], TensorBoardLogger):
+                dirpath = self.loggers[0].log_dir
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:2211]: so the profiler output path and checkpoint path will not be the same when using `Tensorboard`.

File: src/pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -2207,11 +2206,11 @@ def model(self, model: torch.nn.Module) -> None:
 
     @property
     def log_dir(self) -> Optional[str]:
-        if len(self.loggers) == 1:
-            if isinstance(self.logger, TensorBoardLogger):
-                dirpath = self.logger.log_dir
+        if len(self.loggers) > 0:
+            if isinstance(self.loggers[0], TensorBoardLogger):
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:2210]: can we add this info inside profiler docs?

File: src/pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (tshu-w) [reply src/pytorch_lightning/trainer/trainer.py:2211]: Yes, but this is a pre-existing issue that will be addressed in #14188.

File: src/pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (tshu-w) [reply src/pytorch_lightning/trainer/trainer.py:301]: I have added some basic info https://github.com/Lightning-AI/lightning/pull/14325/commits/bff87a0d082d346da5d029e1dea7e30a89c670f9, let me know how to polish it.

File: src/pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (tshu-w) [reply src/pytorch_lightning/trainer/trainer.py:2210]: I find the profilers arguments doc has mentioned where to store profiles If `dirpath` is `None` but `filename` is present: https://github.com/Lightning-AI/lightning/blob/master/src/pytorch_lightning/profilers/simple.py#L46-L48

I'm not sure where should we add. At https://github.com/Lightning-AI/lightning/blob/master/docs/source-pytorch/tuning/profiler_basic.rst?plain=1#L100?

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py:2210]: yes, seems reasonable

File: src/pytorch_lightning/CHANGELOG.md [hunk 6]
```diff
@@ -30,6 +30,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Enabled `on_before_batch_transfer` for `DPStrategy` and `IPUAccelerator` ([14023](https://github.com/Lightning-AI/lightning/pull/14023))
 
 
+- Save checkpoints and profiler output to the first logger ([14325](https://github.com/Lightning-AI/lightning/pull/14325))
```
Comment (awaelchli) [inline src/pytorch_lightning/CHANGELOG.md:33]:
```suggestion
- When using multiple loggers, by default checkpoints and profiler output now get saved to the log dir of the first logger in the list ([14325](https://github.com/Lightning-AI/lightning/pull/14325))
```

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:2211]: Yes, perfect!

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 7]
```diff
@@ -591,15 +590,10 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
...
-            if len(trainer.loggers) == 1:
-                assert trainer.logger is not None
-                save_dir = trainer.logger.save_dir or trainer.default_root_dir
-            else:
-                save_dir = trainer.default_root_dir
-
-            name = _name(trainer.loggers)
-            version = _version(trainer.loggers)
+        elif len(trainer.loggers) > 0:
+            save_dir = trainer.loggers[0].save_dir or trainer.default_root_dir
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:594]: This implementation doesn't allow for `""` (empty string) as a valid `save_dir`. I think it should be (the local directory). This would have been a bug in the previous code too

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -297,9 +297,8 @@ def __init__(
 
             logger: Logger (or iterable collection of loggers) for experiment tracking. A ``True`` value uses
                 the default ``TensorBoardLogger``. ``False`` will disable logging. If multiple loggers are
-                provided and the `save_dir` property of that logger is not set, local files (checkpoints,
-                profiler traces, etc.) are saved in ``default_root_dir`` rather than in the ``log_dir`` of any
-                of the individual loggers.
+                provided, local files (checkpoints, profiler traces, etc.) are saved in the ``log_dir`` of
```
[the remark this answers is not in the document]
Comment (tshu-w) [reply src/pytorch_lightning/trainer/trainer.py:300]: Not really, but specifically `save_dir/name/version`.
https://github.com/Lightning-AI/lightning/blob/31240d39df0576b931aa1b8cd9f1e16d55394b3a/src/pytorch_lightning/callbacks/model_checkpoint.py#L593-L599

 So, in my opinion, we can just to keep `log_dir` for now and leave it for https://github.com/Lightning-AI/lightning/issues/14188

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 9]
```diff
@@ -591,15 +590,10 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
         if trainer._weights_save_path_internal != trainer.default_root_dir:
```
[the remark this answers is not in the document]
Comment (tshu-w) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:590]: same as https://github.com/Lightning-AI/lightning/pull/14325#discussion_r952487498, let's leave the concept of clarification for later #14188, the purpose of this commit is to save checkpoints and profiler output to the first logger.

File: src/pytorch_lightning/callbacks/model_checkpoint.py (hunk 7 above)
Comment (tshu-w) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:598]: I make a new commit to make sure `save_dir` can be empty str.
